### PR TITLE
Fix bmpcodec 16bpp decoding

### PR DESCRIPTION
**Pull request by hughbe**
Fixes https://github.com/mono/libgdiplus/issues/189
Fixes https://github.com/mono/libgdiplus/issues/157

### CODE REVIEW

File: src/bmpcodec.c [hunk 1]
```diff
@@ -827,6 +839,34 @@ gdip_read_BITMAPINFOHEADER (void *pointer, ImageSource source, BITMAPV5HEADER *b
...
+	bmi->bV5GreenMask = (data_read[3]<<24 | data_read[2]<<16 | data_read[1]<<8 | data_read[0]);
+
+	dw = 0;
+	size_read = gdip_read_bmp_data (pointer, data_read, size, source);
+	if (size_read < size)
+		return OutOfMemory;
+	bmi->bV5BlueMask = (data_read[3]<<24 | data_read[2]<<16 | data_read[1]<<8 | data_read[0]);
+
+	return Ok;
 }
```
Comment (filipnavara) [inline src/bmpcodec.c]: This doesn't sound right. I don't think the masks are even present there with BITMAPINFOHEADER when `biCompression` is `BI_RGB`. In that case it should use the default 16 bits per pixel values:

```
bmi->red_mask = 0x7C00;
bmi->green_mask = 0x3E0;
bmi->blue_mask = 0x1F;
```

File: tests/testtext [hunk 2]
```diff
@@ -0,0 +1,210 @@
+#! /bin/sh
```
Comment (filipnavara) [inline tests/testtext]: This is now .gitignored right? Shouldn't be part of the commit. :-)

File: src/bmpcodec.c (hunk 1 above)
Comment (hughbe) [reply src/bmpcodec.c]: The cause of this issue was https://github.com/mono/libgdiplus/issues/285 which needs to be fixed (I'll do so separately)

Addressed this feedback though

File: src/bmpcodec.c [hunk 3]
```diff
@@ -738,7 +750,7 @@ gdip_read_BITMAPINFOHEADER (void *pointer, ImageSource source, BITMAPV5HEADER *b
 		bmi->bV5ClrUsed = 0;
 		bmi->bV5ClrImportant = 0;
 
-		return Ok;
+		goto fill_rgb_masks;
```
Comment (filipnavara) [inline src/bmpcodec.c]: This looks kind of useless. There will never be masks since there's no compression and hence no BI_BITFIELDS (`bV5Compression` should be set to `BI_RGB` for clarity).  And it will never be 16-bit since the allowed values for bcBitCount are only 1, 4, 8, and 24 (https://msdn.microsoft.com/en-us/library/windows/desktop/dd183372(v=vs.85).aspx).

File: src/bmpcodec.c [hunk 4]
```diff
@@ -827,6 +839,41 @@ gdip_read_BITMAPINFOHEADER (void *pointer, ImageSource source, BITMAPV5HEADER *b
...
+		bmi->bV5BlueMask = 0;
+		return Ok;
+	}
+	if (bmi->bV5Compression != BI_BITFIELDS) {
+		// Non-compressed 16bpp bitmaps are always RGB555.
+		bmi->bV5RedMask = 0x7C00;
+		bmi->bV5GreenMask = 0x3E0;
+		bmi->bV5BlueMask = 0x1F;
+		return Ok;
+	}
```
Comment (filipnavara) [inline src/bmpcodec.c]: These two block could probably be moved up to the "goto fill_rgb_masks" point and then the mask reading code could be shared instead of being duplicated. In fact, I think the `if (bmi->bV5BitCount != 16)` block is unnecessary.

File: src/bmpcodec.c [hunk 5]
```diff
@@ -784,8 +796,29 @@ gdip_read_BITMAPINFOHEADER (void *pointer, ImageSource source, BITMAPV5HEADER *b
...
+				return OutOfMemory;
+			bmi->bV5GreenMask = (data_read[3]<<24 | data_read[2]<<16 | data_read[1]<<8 | data_read[0]);
+
+			dw = 0;
+			size_read = gdip_read_bmp_data (pointer, data_read, size, source);
+			if (size_read < size)
+				return OutOfMemory;
+			bmi->bV5BlueMask = (data_read[3]<<24 | data_read[2]<<16 | data_read[1]<<8 | data_read[0]);
+		}
+
```
Comment (filipnavara) [inline src/bmpcodec.c]: Any reason why you don't fall through to the code below? Also, now the implicit BI_RGB masks from previous change are missing...

Comment (filipnavara) [reply src/bmpcodec.c]: For the record I don't mind the code duplication if it makes the code more readable. It's matter of personal opinion and I find your version little more readable even though it's a bit verbose... but the BI_RGB part has to be fixed in either case.
